Add email time of day preference

elliotthilaire opened the pull request:
Closes https://github.com/codetriage/codetriage/issues/244

This adds an email time of day setting on the users settings page.

<img width="1363" alt="screen shot 2017-02-10 at 2 44 41 pm" src="https://cloud.githubusercontent.com/assets/3443017/22831069/d836a37a-efa0-11e6-98e5-65e51438bd4b.png">

<img width="1363" alt="screen shot 2017-02-10 at 2 42 08 pm" src="https://cloud.githubusercontent.com/assets/3443017/22831025/adf743b2-efa0-11e6-9d18-177fd3c2a15f.png">
 
<img width="1363" alt="screen shot 2017-02-10 at 2 42 37 pm" src="https://cloud.githubusercontent.com/assets/3443017/22831067/d62eaffa-efa0-11e6-9cd3-ec05ff60d626.png">

The Heroku schedular can be set up to run hourly and `send_daily_triage!` will skip sending the triage if it is before the users set preference.

I can set a default email time of day. How would you like to do this?
1. A default set in the code
2. A rake task to update users

This PR includes extra changes in schema.rb. These changes come from running db:migrate on afresh setup. Details here: https://github.com/codetriage/codetriage/pull/526

Happy to take on feedback and make further changes.

elliotthilaire commented:
Hello @schneems,  @prathamesh-sonpatki 

A further comment on this. I have not tested this on a live server. I looked at setting up my own Heroku instance and running it for a few days before submitting a pull request.

This is what stopped me:
1. I don't know what/when scheduled tasks are set up live, so I would be guessing about the configuration.
2. I'm not sure if I can it relying only on free services/addons, redis, email, etc.
3. newrelic.yml and skylight.yml have hardcoded keys which might interfere.

Do you have a staging setup this can be tested on?
Would you be willing to assist me in setting up my own?

MonikaMahanthappa commented on db/schema.rb:
```diff
@@ -73,8 +73,8 @@
-    t.datetime "updated_at",                           null: false
+    t.datetime "created_at"
+    t.datetime "updated_at"
     t.integer  "repo_subscription_id"
```
The default database constraints are getting removed. Unless it is intentional, would suggest that you do not make any changes not related to your story in this PR. Getting it back to as it is on master. 

This issue might have caused due to some Postgres version difference.

MonikaMahanthappa commented on test/unit/user_test.rb:
```diff
@@ -93,6 +93,91 @@ class UserTest < ActiveSupport::TestCase
+    end
+  end
+
+  test 'send_daily_triage! can be scheduled to run multiple times a day' do
```
I think that this test is not required/redundant. It is already covered in the above tests.

elliotthilaire replied:
Thanks @MonikaMahanthappa. The tests above you refer too only run the send_daily_triage! method once.

This test was to ensure that if we call send_daily_triage! multiple times after the preferred time we won't send multiple emails. Since this is the responsibility of EmailDecider I agree it is a redundant.

MonikaMahanthappa commented on test/unit/user_test.rb:
```diff
@@ -93,6 +93,91 @@ class UserTest < ActiveSupport::TestCase
+    issue.state           = 'open'
+    issue.html_url        = "http://schneems.com"
+    issue.number          = (repo.issues.last.try(:number) || 0) + 1
+    issue.save
```
It is not required to create the issue with all attributes set as this test is not particularly testing any of those values. Eg, It can be `repo.issues.create(title: 'Foo Bar')` or `issue = repo.issues.new(title: 'Foo Bar'); issue.save`

elliotthilaire replied:
Thanks @MonikaMahanthappa. I'll take a look to see how it could be simplified. I'm not sure if I will run into issues where some attributes have to be set before they are saved.
I duplicated the existing tests on send_daily_triage! which have this setup.

MonikaMahanthappa commented on app/helpers/users_helper.rb:
```diff
@@ -0,0 +1,5 @@
+module UsersHelper
+  def available_times_of_day
+    (0...24).map { |hour| Time.utc(2000, 1, 1, hour, 00, 0) }
```
To make it easier for the user, I suggest that we take the user timezone also into consideration.

elliotthilaire replied:
Thanks @MonikaMahanthappa, I agree.  I wanted to create the simplest PR possible so I left out user time zones. 

If we present the times as UTC and save the times as UTC we have a working feature and a solid base to build user time zones on.

Do you think this approach is ok?

elliotthilaire commented:
I've updated this with feedback from @MonikaMahanthappa. Thanks Monika.

I've also added a default time of 5pm UTC.
This is so that the scheduled task can run hourly without a change for when people receive their emails.
This default was chosen based on when I have been getting my emails.

nateberkopec commented:
> I have not tested this on a live server. 

@elliotthilaire can you rebase this? I think review apps are set up now, so you should be able to test this live if you rebase and push.

elliotthilaire commented:
Hi @nateberkopec,

I've rebased and pushed. 
Heroku CI has failed but I'm unable to see why.

schneems commented:
If travis is happy then I am too. There's some wonkyness going on with the Heroku suite. I migrated the app to an org and then back again and Heroku did not like that at all. 

If you can address the issues Nate touched on I think we can get this over the line.

elliotthilaire commented:
I am working through conflicts introduced by the refactor:
https://github.com/codetriage/codetriage/pull/554/files

I am planning on adding display of local timezone using gem 'local_time'

schneems commented:
Builds are working. What else is left?

> I am planning on adding display of local timezone using gem 'local_time'

Is that it?

nateberkopec commented:
@schneems I'd like to another code review round before this gets merged if that's OK w/you

elliotthilaire commented on config/initializers/new_framework_defaults.rb:
```diff
@@ -20,3 +20,6 @@
 ActiveSupport.halt_callback_chains_on_return_false = true
+
+# Time columns will become time zone aware in Rails 5.1
+Rails.application.config.active_record.time_zone_aware_types = [:datetime, :time]
```
I think this can go now that we are Rails 5.1

elliotthilaire commented:
Builds are working. What else is left?

       I am planning on adding display of local timezone using gem 'local_time'

   Is that it?

@schneems Yes.

gem 'local_time' does not work with our version of turbolinks.
Pointing gem file to the local_time 2.0 branch in github fixes this.
https://github.com/basecamp/local_time/pull/54

I can't use local_time as a drop in on top of the select box options because the display values can't include html tags. I need to figure out a nice way to do this.

schneems commented:
Looks like we can use https://github.com/basecamp/local_time/pull/55

Can you try out

```
gem 'local_time', git: 'https://github.com/basecamp/local_time', branch: '2-0'
```

elliotthilaire commented:
> Looks like we can use basecamp/local_time#55

Yep. That's what I did to solve the first issue.

The second issue is that you cannot do `<option value="..."><%= local_time(time) %></option>`
So I need to find another way.

elliotthilaire commented:
I think this is done.

The server doesn't know about the users timezone. The times are always sent to and from in UTC, and stored in UTC. If javascript is enabled. It will replaced the display of the times to the local time.

<img width="769" alt="screen shot 2017-06-03 at 4 22 50 am" src="https://cloud.githubusercontent.com/assets/3443017/26739315/acbb361c-4814-11e7-84cc-954919cc0871.png">

<img width="770" alt="screen shot 2017-06-03 at 4 23 49 am" src="https://cloud.githubusercontent.com/assets/3443017/26739325/b6021d1c-4814-11e7-9c38-92e3fb04a159.png">

The order of the select box does not change. They are ordered by value, not text. I don't think this should be a blocker for this PR. It might be more suitable for someone with better JS fu than I.  I did have a look into reordering but wasn't happy with the solutions I could come up with.
 
<img width="592" alt="screen shot 2017-06-03 at 4 31 35 am" src="https://cloud.githubusercontent.com/assets/3443017/26739551/9a109e7a-4815-11e7-89a0-9df2f5bfc1af.png">

elliotthilaire commented:
Happy for another round of review but am confident if you merge.
Once merged, you can change Heroku to run the scheduled task each hour.

elliotthilaire commented:
I've rebased and fixed conflicts. Then squashed to cleanup history.

nateberkopec commented on app/jobs/send_daily_triage_email_job.rb:
```diff
@@ -28,4 +28,14 @@ def skip_daily_email?(user)
+
+  DEFAULT_EMAIL_TIME_OF_DAY = Time.utc(2000, 1, 1, 17, 00, 0)
+
+  def before_email_time_of_day?(user)
```
Comparing times as strings seems weird. Why not just compare the times directly?

elliotthilaire replied:
The time column stored the time as well as the data 2000-1-1. 
```
irb(main):002:0> User.first.email_time_of_day.utc
 => 2000-01-01 17:00:00 UTC
```
This is comparing the time it is now for today, against the time that is set in the database (ignoring the data part).

If you know of a method to handle this, I'm open for suggestions.

It's why you also see this further down
```
Time.utc(now.year, now.month, now.day, time.hour, time.min, time.sec)
```

nateberkopec replied:
Let's just convert time.current:

```ruby
Time.current.change(:month => 1, :day => 1, :year => 2000)
```

elliotthilaire replied:
Done. 

I thought this might not work in the case that a different date ever got stored by accident. e.g. via console or database manipulation. 

But this can't happen. The 2000-1-1 date is never stored in the table. It gets added when it's turned into a ruby object.

```
\d users
...
| email_time_of_day      | time without time zone      |                                                                  |
...
```

```
triage_development> select email_time_of_day from users;
+---------------------+
| email_time_of_day   |
|---------------------|
| 15:31:01.986967     |
+---------------------+
```

nateberkopec commented on test/jobs/send_daily_triage_email_job_test.rb:
```diff
@@ -8,20 +8,64 @@ def setup
 
   test 'does not deliver if email should be skipped' do
     def @user.issue_assignments_to_deliver; IssueAssignment.all.limit(1); end
+    def @job.before_email_time_of_day?(*); false; end
```
no need for this line because skip_daily_email? returning true will short circuit the conditional

elliotthilaire replied:
That's true. 

If I remove this line and then and `skip_daily_email?` develops a bug.
Then this will pass with a false positive depending on the time of day it is run.

Correct me if I'm wrong.

elliotthilaire replied:
I am wrong, because `skip_daily_email?` is mocked.

nateberkopec commented on test/jobs/send_daily_triage_email_job_test.rb:
```diff
@@ -8,20 +8,64 @@ def setup
+  test 'when email_time_of_day not set, it delivers after default time of day' do
+    def @user.issue_assignments_to_deliver; IssueAssignment.all.limit(1); end
+
+    Timecop.freeze time_preference_for_today(SendDailyTriageEmailJob::DEFAULT_EMAIL_TIME_OF_DAY) + 1.hour do
```
Rather than use a new testing library, we can just use [Minitest stubs](https://github.com/seattlerb/minitest/blob/master/lib/minitest/mock.rb#L212).

Time.stub(:now, your_time) do ...

nateberkopec commented:
Thanks for seeing this through!

elliotthilaire commented:
> Thanks for seeing this through!

@nateberkopec, No problem. 

I've left comments for 2/3 of the requested changes.

elliotthilaire commented:
I've rebased and changed my refutes to assert_nots based on. https://github.com/codetriage/codetriage/commit/789a6a74206728486ea7d05172c772976b594443

nateberkopec commented:
> I've rebased and changed my refutes to assert_nots based on

Thank you! @schneems told me to do that.

nateberkopec commented:
One last comment about code, but otherwise LGTM. Let me figure out how to get this on a review app...

schneems commented:
🎉 this is amazing. I want to babysit this deploy a bit, so i'm going to wait until next week to merge. Thanks again for all the work!

elliotthilaire commented:
> 🎉 this is amazing. I want to babysit this deploy a bit, so i'm going to wait until next week to merge. Thanks again for all the work!

No problem. If you get a PR deploy running I'm happy to test. Let me know if there's anything else that pops up that needs attention before merge.

schneems commented:
Patch is merged. Thanks again for all your hard work!

Should be deployed shortly.

elliotthilaire commented:
@schneems Is everything ok with this?

I haven't received an email for the last couple of days.

I've set my preference to be 6am UTC.  The last daily email I received was on the 21st at the unchanged time (5:40~5:50)

schneems commented:
I think I ran out of sendgrid credits

elliotthilaire commented:
Ok. That's fine. I was concerned I broke something. If there's an issue with the feature; I'll fix it.

schneems commented:
Also I forgot to set the scheduler to run hourly instead of once a day. I just fixed that.

schneems commented:
There's a bug somewhere. I got this report. I'm trmporarilly disabling emails https://twitter.com/rx14_chibi/status/878715739011776512

elliotthilaire commented:
@schneems Sorry about this. I'm looking into it.

From the screenshot. It's not sending emails out every hour. It send out 7 consecutive emails.
I think `user.days_since_last_clicked` is working on a different day to `before_email_time_of_day?` so there is an overlap.

`days_since_last_clicked` is using Time.now (which returns the time in timezone of the server)
The new feature uses `Time.current` (which returns the time in UTC)

Could you confirm what timezone the server is at? I assume it is about a 7 hour difference to UTC?

I think we can fix this by changing user to use `Time.current` instead of `Time.now` so it it timezone agnostic.

I think this is also related to the random test failure https://github.com/codetriage/codetriage/issues/541
